crypto/rsa/rsa_gen.c

```diff
@@ -186,10 +188,18 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, int primes, BIGNUM *e_value,
             }
             if (!BN_sub(r2, prime, BN_value_one()))
                 goto err;
-            if (!BN_gcd(r1, r2, rsa->e, ctx))
-                goto err;
-            if (BN_is_one(r1))
+            if (BN_mod_inverse(r1, r2, rsa->e, ctx)) {
+               /* GCD == 1 since inverse exists */
                 break;
+            }
+            error = ERR_peek_last_error();
+            if (ERR_GET_LIB(error) == ERR_LIB_BN
+                && ERR_GET_REASON(error) == BN_R_NO_INVERSE) {
+                    /* GCD != 1 */
+                    ERR_clear_error();
```

Comment:
I wonder whether `ERR_set_mark()` before the BN_mod_inverse() call, and `ERR_pop_to_mark()` here might be better?

Reply:
As far as I understood, set/pop mark cannot be used to check for a specific error. In our case we want to distinguish BN_R_NO_INVERSE from other errors like out-of-memory. Former should cause a retry using a different prime number, staying inside the for-loop. Latter should abort to avoid endless loops.

Reply:
They are not intended as a replacement for the `ERR_peek_last_error()` call. You still need to do that to check what the last error code was and therefor look at `BN_R_NO_INVERSE`. Rather they are a replacement for `ERR_clear_error()`. That function clears *all* errors on the error queue. By using `ERR_set_mark()` and `ERR_pop_to_mark()` you can clear only those errors added by the `BN_mod_inverse()` call.